firewood/src/merkle/stream.rs

```diff
@@ -1,47 +1,228 @@
+        /// we pop it off the stack and go back up to its parent.
+        branch_iter_stack: Vec<BranchIterator>,
     },
 }
 
-impl IteratorState<'_> {
+impl IteratorState {
     fn new() -> Self {
-        Self::StartAtKey(vec![].into_boxed_slice())
+        Self::Uninitialized(vec![].into_boxed_slice())
     }
 
     fn with_key(key: Key) -> Self {
-        Self::StartAtKey(key)
+        Self::Uninitialized(key)
     }
 }
 
 /// A MerkleKeyValueStream iterates over keys/values for a merkle trie.
 pub struct MerkleKeyValueStream<'a, S, T> {
-    key_state: IteratorState<'a>,
+    key_state: IteratorState,
     merkle_root: DiskAddress,
     merkle: &'a Merkle<S, T>,
 }
 
+/// Returns the state of an iterator over merkle, which has the given
+/// root_node, after it's been initialized. Iteration starts at the given key.
+fn get_iterator_intial_state<S: ShaleStore<Node> + Send + Sync, T>(
+    merkle: &Merkle<S, T>,
+    root_node: DiskAddress,
+    key: &[u8],
+) -> Result<IteratorState, api::Error> {
+    let root_node = merkle
+        .get_node(root_node)
+        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+    let mut branch_iter_stack: Vec<BranchIterator> = vec![];
+
+    // (disk address, index) for each node we visit along the path to
+    // [key], where [index] is the next nibble in the key after the node
+    // at [disk address].
+    let mut path_to_key = vec![];
+
+    // Populate [path_to_key].
+    merkle
+        .get_node_by_key_with_callbacks(
+            root_node,
+            key,
+            |node_addr, i| path_to_key.push((node_addr, i)),
+            |_, _| {},
+        )
+        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+    // Get each node in [path_to_key]. Mark the last node as being the last
+    // so we can use that information in the while loop below.
+    let path_to_key = path_to_key
+        .into_iter()
+        .rev()
+        .enumerate()
+        .rev()
+        .map(|(i, (node, pos))| merkle.get_node(node).map(|node| (i == 0, (node, pos))));
+
+    let mut matched_key_nibbles = vec![];
+
+    for node in path_to_key {
+        let (is_last, (node, pos)) = match node {
+            Ok(result) => result,
+            Err(e) => return Err(api::Error::InternalError(Box::new(e))),
+        };
+
+        match node.inner() {
+            NodeType::Leaf(_) => (),
+            NodeType::Branch(branch) if is_last => {
+                // This is the last node in the path to [key].
+                // Figure out whether to start iterating over this node's
+                // children at [pos] or [pos + 1].
+
+                // Get the child at [pos], if any.
+                let Some(child) = branch.children.get(pos as usize) else {
+                    // This should never happen -- [pos] should never be OOB.
+                    return Err(api::Error::InternalError(Box::new(
+                        api::Error::ChildNotFound,
+                    )));
+                };
```

Comment:
This is a valid case for a panic, as the code prevents this from ever happening.

See https://blog.burntsushi.net/unwrap/#so-one-should-never-use-unwrap-or-expect

We have a rule for prohibiting unwrap() but not expect() exactly for this case.
```suggestion
                // Get the child at [pos], if any.
                let child = branch.children.get(pos as usize).expect("pos cannot be OOB");
```